Design/Instrument-2.0/Instrument_Ray_Tracer_2.md

```diff
@@ -0,0 +1,68 @@
+
+#### Classless Implementation
+Another idea could be to make the methods of `InstrumentRayTracer 2.0` a set of free functions. Each of the functions would need to take in a `ComponentInfo` and any other variables required to carry out the correct procedure. This approach would definitely eliminate the passing around of the `InstrumentRayTracer` object that currently happens. Also, there are not many variables that would need to be passed to each of the methods meaning calls to the functions would remain largely the same.
+
+## Usage of `ComponentInfo`
+If all goes well with the plan to use `ComponentInfo` where possible, it is very likely that the code to use `InstrumentRayTracer 2.0` will be very similar to the current method of using it. Possible methods from `ComponentInfo` that could be used in `InstrumentRayTracer 2.0` might include:
+
+ * `componentInfo.getSource()` in `void InstrumentRayTracer::trace(const V3D &dir) const`
+ * `componentInfo.getSample()` in `void InstrumentRayTracer::traceFromSample(const V3D &dir) const`
+
+It should not be too difficult to reuse the code from `ComponentInfo` as it should just be a case of making sure the `ComponentInfo` class is accessible from `InstrumentRayTracer 2.0`.
+
+## Rolling out the Changes
+Before rolling out any changes, it is probably a good idea to develop a set of tests exclusively for this new class. Some tests (especially for methods that do not make calls to the legacy API) could probably be based on the existing tests with a few alterations. 
+
+#### Changes to the Code Base
+The best way to completely move away from using `InstrumentRayTracer` (and not completely break everything) would be to start with a selection of "testing files" that currently use the `InstrumentRayTracer` e.g `Peak` and begin to change them one by one. Once all of those files are in working order it would provide enough confidence in the new implementation and all the other files can then be updated.
+
+#### Performance Issues
+To ensure that the new implementation performs just as well as the existing version, benchmarking tools could be used to record the current performance and this can be used as a target for the new implementation. It does not seem that there would be any major changes to performance since much of the code should remain somewhat similar. If anything, using `ComponentInfo` should reduce the number of function calls being made. 
+
+## Conclusion
+Based on the options outlined above, it seems that the best approach may be to make `InstrumentRayTracer 2.0` a class based implementation. The constructor should take in a copy of a `ComponentInfo` object. The reasoning behind this is that the current implementation is class based and keeping the new version familiar in design might be good for end users. Also, there should not be any extra overhead in terms of performance since the implementation strategy is similar to the previous version.
```

Comment:
I disagree with this. Say you wanted to trace a ray. With the class approach I need to write 
```cpp
InstrumentRayTracer instrumentRayTracer(ws.componentInfo());
instrumentRayTracer.trace(V3D{0,0,0});
auto links = instrumentRayTracer.getResults();
```
End users (read Mantid Developers) would likely find the following one-liner more convenient,
```cpp
auto links = InstrumentRayTracer::trace(ws.componentInfo(), V3D{0,0,0});
```
Note that these are slightly different considerations from our Python API, which we would put more effort into preserving backwards compatibility. That restriction does not exist for our C++ API especially for internals like the `InstrumentRayTracer`

For performance reasons, I also think the latter would be better.
1. No copy of `ComponentInfo` required, even though the COW mechanism makes copies of `ComponentInfo` relatively cheap, this is unnecessary.
1. `trace` internally calls `reset` on on `Track` (`m_resultsTrack`) every call, it would likely be just as efficient to construct a new `Track` every time.
1. `getResults` returns by value and copies the links into a new temporary before resetting/clearing it's internal state.

Reply:
I see, I think you are right. I think my reasoning was more based on what we would've done for the Python side. It is probably best to use the classless implementation if it's easier for developers and there are performance benefits.